Core REST API: Add permission callback to delete_jitm_message endpoint

kbrown9 opened the pull request:
See #16539 

#### Changes proposed in this Pull Request:
* Only users with the 'read' capability should be able to dismiss JITMs. Add a new permission callback to the `delete_jitm_message` endpoint that restricts dismissals to users that have the 'read' capability.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This changes an existing part of Jetpack.

#### Testing instructions:

##### Test Site
* Must have Jetpack active and connected.
* Must display JITMs (that is, they haven't been dismissed.)

##### Test Steps
1. Confirm that JITMs can be successfully dismissed in wp-admin:
   a. Navigate to the Jetpack dashboard. Dismiss the JITM.
   b. Reload the page and confirm that the JITM is not displayed.
   c. Confirm that the `hide_jitm` option was updated using the command shown below. The dismissed JITM should be in the `hide_jitm` array.
          
       wp jetpack options get hide_jitm

2. Confirm that JITMs can be successfully dismissed in Calypso:
   a. Navigate to a Calypso page that displays a JITM, on the themes upload page.
   b. Dismiss the JITM.
   c. Reload the page and confirm that the JITM not displayed.

(NOTE: I haven't been able to see a JITM in Calypso, so I haven't been able to test this.)

3. Attempt to make a post request to the JITM endpoint: 

      `wp-json/jetpack/v4/jitm`

    a. The post request body should be a JSON request with the keys 'id' and 'feature_class'. For example:
         ``` { "id": "some_id", "feature_class": "some_feature_class" }```

    b. The POST request should be rejected with a 401 error. 

#### Proposed changelog entry for your changes:
* tbd

jeherve commented:
@kbrown9 Should we close now that we've merged #16540?

kraftbj commented:
I'd rather there be an explicit permissions check. My patch was meant mainly to unblock Travis with the intent of coming back and checking each endpoint.

I feels incorrect to have the jitm endpoints be open, so very open to this patch (post-rebase).

kbrown9 commented:
I'd also still like to have the explicit permission check on this endpoint. My blocker on this was testing with Calypso. I'm not sure how JITM dismissals in Calypso are handled, so I need to make sure this change doesn't affect them.

kbrown9 committed:
Core REST API: Add permission callback to delete jitm endpoint

Only users with the 'read' capability should be able to dismiss JITMs. Add
a new permission callback to the delete_jitm_message endpoint that restricts
dismissals to users that have the 'read' capability.

dereksmart reviewed (dismissed):
LGTM

jeherve reviewed (dismissed):
I didn't get the changes right away, and I think it may have been because of this:
https://github.com/Automattic/jetpack/blob/1a6f86724c63bf037e6751ab4b7c91ee3dd57f12/packages/jitm/src/class-jitm.php#L24

Do you think you could update this? Maybe we can use today's date for now, if keeping this up to date for each package version change proves difficult?

kbrown9 committed:
JITM: Update the package version in the JITM class.

kbrown9 commented:
> I didn't get the changes right away, and I think it may have been because of this:
> https://github.com/Automattic/jetpack/blob/1a6f86724c63bf037e6751ab4b7c91ee3dd57f12/packages/jitm/src/class-jitm.php#L24
> 
> Do you think you could update this? Maybe we can use today's date for now, if keeping this up to date for each package version change proves difficult?

Thanks for the review! I updated the package version in commit 10ab918.
